agent/hcp/telemetry_provider.go

```diff
@@ -46,114 +48,138 @@ type hcpProviderImpl struct {
-	refreshInterval := telemetryCfg.RefreshConfig.RefreshInterval
-	// refreshInterval must be greater than 0, otherwise time.Ticker panics.
-	if refreshInterval <= 0 {
-		return nil, fmt.Errorf("invalid refresh interval: %d", refreshInterval)
-	}
-
-	cfg := &dynamicConfig{
-		Endpoint:        telemetryCfg.MetricsConfig.Endpoint,
-		Labels:          telemetryCfg.MetricsConfig.Labels,
-		Filters:         telemetryCfg.MetricsConfig.Filters,
-		RefreshInterval: refreshInterval,
+// defaultDisabledCfg disables metric collection and contains default config values.
+func defaultDisabledCfg() *dynamicConfig {
+	return &dynamicConfig{
+		labels:          map[string]string{},
+		filters:         client.DefaultMetricFilters,
+		refreshInterval: defaultTelemetryConfigRefreshInterval,
+		endpoint:        nil,
+		disabled:        true,
 	}
+}
 
-	t := &hcpProviderImpl{
-		cfg:       cfg,
+// NewHCPProvider initializes and starts a HCP Telemetry provider.
+func NewHCPProvider(ctx context.Context, hcpClient client.Client) *hcpProviderImpl {
+	h := &hcpProviderImpl{
+		// Initialize with default config values.
+		cfg:       defaultDisabledCfg(),
 		hcpClient: hcpClient,
 	}
 
-	go t.run(ctx, refreshInterval)
+	go h.run(ctx)
 
-	return t, nil
+	return h
 }
 
 // run continously checks for updates to the telemetry configuration by making a request to HCP.
-func (h *hcpProviderImpl) run(ctx context.Context, refreshInterval time.Duration) {
-	ticker := time.NewTicker(refreshInterval)
+func (h *hcpProviderImpl) run(ctx context.Context) {
+	// Try to initialize config once before starting periodic fetch.
+	h.updateConfig(ctx)
+
+	ticker := time.NewTicker(h.cfg.refreshInterval)
 	defer ticker.Stop()
 	for {
 		select {
 		case <-ticker.C:
-			if newCfg := h.getUpdate(ctx); newCfg != nil {
-				ticker.Reset(newCfg.RefreshInterval)
+			if newRefreshInterval := h.updateConfig(ctx); newRefreshInterval > 0 {
+				ticker.Reset(newRefreshInterval)
 			}
 		case <-ctx.Done():
 			return
 		}
 	}
 }
 
-// getUpdate makes a HTTP request to HCP to return a new metrics configuration
-// and updates the hcpProviderImpl.
-func (h *hcpProviderImpl) getUpdate(ctx context.Context) *dynamicConfig {
+// updateConfig makes a HTTP request to HCP to update metrics configuration held in the provider.
+func (h *hcpProviderImpl) updateConfig(ctx context.Context) time.Duration {
 	logger := hclog.FromContext(ctx).Named("telemetry_config_provider")
 
 	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
 	defer cancel()
 
 	telemetryCfg, err := h.hcpClient.FetchTelemetryConfig(ctx)
 	if err != nil {
+		// Only disable metrics on 404 or 401 to handle the case of an unlinked cluster.
+		// For other errors such as 5XX ones, we continue metrics collection, as these are potentially transient server-side errors.
+		apiErr, ok := err.(*runtime.APIError)
+		if ok && (apiErr.IsCode(404) || apiErr.IsCode(401)) {
+			disabledMetricsCfg := defaultDisabledCfg()
+			h.modifyDynamicCfg(disabledMetricsCfg)
```

Comment:
@jjti I emit a success metric in this case, since it's not really a failure?

Would it be better to emit a different custom metric, e.g. `internalMetricRefreshDisabled []string = []string{"hcp", "telemetry_config_provider", "refresh", "disabled"}`

My reasoning is that we should treat it as a succesful response and makes it easier to calculate success rate. Otherwise we have to know about this edge case and consider it.

Reply:
It's up to you here, either work. One other option is to use `metrics.IncrCounterWithLabels([]string{"hcp", "telemetry_config_provider", "refresh"}, 1, []metrics.Label{Name:"result", Value:"skipped"})`, etc. So put the result in a label so there's just one counter, but multiple active time series, for the refreshing

You can see a bit of "use the labels to delineate result" here in this (beta) metric in consul that's emitted from all rpc endpoints: https://developer.hashicorp.com/consul/docs/agent/telemetry#server-workload

I prefer it personally to the rest of the Consul metrics that are like `consul.cache.fetch_success` vs `consul.cache.fetch_error` etc

Reply:
I don't feel strongly either way but I feel like a label is a bit odd. I think if we want to explicitly measure the case of 4xx vs 5xx vs legitmately successful updates we could do it with different metrics. Wouldn't die on a hill and will leave it up to the observability team.

Reply:
I also wouldn't die on this hill. I haven't seen a recommendation for one approach vs other (they result in the same total sample count + cost in a DD or AMP). I slightly prefer the label tho. Eg from an Envoy HTTP response counter:

<img width="624" alt="Screenshot 2023-08-29 at 1 58 01 PM" src="https://github.com/hashicorp/consul/assets/13923102/345f2f1b-a60c-4818-8db8-18341d50888f">

Reply:
After walking through an example, there are reasons why this is a good idea. I will target a follow-up PR to do this since we need to modify all other metrics to follow the same pattern (for consistency_